rust-runtime/aws-smithy-http/src/byte_stream/bytestream_util.rs

```diff
@@ -139,56 +161,90 @@ impl FsBuilder {
         self
     }
 
-    /// Returns a [`ByteStream`](crate::byte_stream::ByteStream) from this builder.
-    /// NOTE: If both [`file`](FsBuilder::file) and [`path`](FsBuilder::path) have been called for this FsBuilder, `build` will
-    /// read from the path specified by [`path`](FsBuilder::path).
+    /// Specify the offset to start reading from (in bytes)
     ///
-    /// # Panics
-    ///
-    /// Panics if neither of the `file` or`path` setters were called.
+    /// When used in conjunction with [`length`](FsBuilder::length), allows for reading a single "chunk" of a file.
+    pub fn offset(mut self, offset: u64) -> Self {
+        self.offset = Some(offset);
+        self
+    }
+
+    /// Returns a [`ByteStream`](crate::byte_stream::ByteStream) from this builder.
     pub async fn build(self) -> Result<ByteStream, Error> {
+        if self.path.is_some() && self.file.is_some() {
+            panic!("The 'file' and 'path' options on an FsBuilder are mutually exclusive but both were set. Please set only one")
+        };
+
         let buffer_size = self.buffer_size;
+        let offset = self.offset.unwrap_or(DEFAULT_OFFSET);
+        // Checking the file length like this does have a cost, but the benefit is that we can
+        // notify users when file/chunk is smaller than expected.
 
-        if let Some(path) = self.path {
-            let path_buf = path.to_path_buf();
-            let file_size = self.file_size.unwrap_or(
-                tokio::fs::metadata(path)
-                    .await
-                    .map_err(|err| Error(err.into()))?
-                    .len(),
-            );
+        let length = match self.length {
+            Some(Length::Exact(length)) => {
+                let file_length = self.get_file_size().await?;
+                if length > file_length.saturating_sub(offset) {
+                    return Err(Error(Box::new(io::Error::new(
+                        io::ErrorKind::UnexpectedEof,
+                        "value Length::Exact was larger than file size minus read offset",
+                    ))));
```

Comment:
Looks like there isn't a unit test exercising this error.

Reply:
good callout, I'll create one